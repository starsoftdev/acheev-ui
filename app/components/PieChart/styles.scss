@import '~styles/settings/settings';
.pieChart {
  position: relative;
  &__svg {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
  &__inner {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }
  &__label {
    color: $medium-dark-gray;
    font-size: $fs-md;
    font-weight: bold;
  }
  &__value {
    font-size: $fs-xl;
    font-weight: bold;
    color: $darker-gray;
  }
  &__pancakeRow {
    margin: $rem-sm 0;
  }
  &__pancake {
    margin: 0 auto;
    border-radius: 50%;
  }
}
