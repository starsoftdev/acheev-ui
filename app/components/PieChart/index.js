// @flow

import React from 'react';
import { PieChart, Pie, Cell } from 'recharts';
import { generate } from 'shortid';
import { maxBy, minBy, get } from 'lodash-es';
import cx from 'classnames';

import './styles.scss';

type Props = {
  data: ?Array<Object>,
  message?: string,
  innerComponent?: any,
  label?: string,
  labelPrefix?: Function,
  valuePostfix?: string,
  palette?: 'blue' | 'green',
  className?: string,
  radius?: number,
  thickness?: number,
  stroke?: string,
  stackForSmall?: boolean,
};

const COLORS: Object = {
  blue: ['#003fa3', '#017fcc', '#11b4e4', '#84d6ef', '#c9eaf6'],
  green: ['#276c07', '#64a62a', '#a0ce67', '#cbe3b1', '#e6f0dd'],
};

const CustomPieChart = ({
  data,
  message = 'No data for selected period',
  innerComponent,
  label,
  labelPrefix,
  valuePostfix,
  palette = 'blue',
  className,
  radius = 112.5,
  thickness = 25,
  stroke = 'white',
  stackForSmall,
}: Props) => {
  if (!data || !data.length) return <h6>{message}</h6>;
  const maxCount = get(maxBy(data, o => o.count), 'count', 0);
  const minCount = get(minBy(data, o => o.count), 'count', 0);
  const minSize = 15;
  const maxSize = 30;
  return (
    <div className={cx('row align-middle align-center', className)}>
      <div className="shrink column">
        <div className="pieChart">
          <PieChart
            className="pieChart__svg"
            width={radius * 2}
            height={radius * 2}
          >
            <Pie
              data={data}
              dataKey="count"
              innerRadius={radius - thickness - 2}
              outerRadius={radius}
            >
              {data.map((entry, index) => (
                <Cell
                  fill={COLORS[palette][index % COLORS[palette].length]}
                  strokeWidth="5"
                  stroke={stroke}
                  key={generate()}
                />
              ))}
            </Pie>
          </PieChart>
          {innerComponent && (
            <div className="pieChart__inner">{innerComponent}</div>
          )}
        </div>
      </div>
      <div
        className={cx('pieChart__legend column', {
          'small-12 medium-expand': stackForSmall,
        })}
      >
        {data.map((entry, index) => {
          const value = entry.count;
          let size;
          if (value === maxCount) {
            size = maxSize;
          } else if (value === minCount) {
            size = minSize;
          } else {
            size = (minSize + maxSize) / 2;
          }
          return (
            <div
              className="pieChart__pancakeRow row align-middle"
              key={generate()}
            >
              <div className="shrink column">
                <div style={{ width: maxSize }}>
                  <div
                    className="pieChart__pancake"
                    style={{
                      backgroundColor:
                        COLORS[palette][index % COLORS[palette].length],
                      width: size,
                      height: size,
                    }}
                  />
                </div>
              </div>
              <div className="column">
                <div className="pieChart__value">
                  {entry.count || 0}
                  {valuePostfix}
                </div>
                <div className="pieChart__label">
                  {labelPrefix && labelPrefix(entry)} {label}
                </div>
              </div>
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default CustomPieChart;
