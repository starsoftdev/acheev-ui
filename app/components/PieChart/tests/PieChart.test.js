import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';

import PieChart from 'components/PieChart';

const data = [
  { name: '10$', price: 10, count: 40 },
  { name: '20$', price: 20, count: 20 },
  { name: '50$', price: 20, count: 20 },
];

const emptyData = [];

describe('<PieChart />', () => {
  it("shouldn't throw error if data is not specified", () => {
    try {
      shallow(<PieChart />);
    } catch (error) {
      expect(error).toBeUndefined();
    }
  });

  it('renders message if data is not provided', () => {
    const component = shallow(<PieChart message="Hey" />);
    expect(component.contains('Hey')).toEqual(true);
  });

  it('renders message if empty array is provided', () => {
    const component = shallow(<PieChart data={emptyData} message="Hey" />);
    expect(component.contains('Hey')).toEqual(true);
  });

  it('renders `message` if provided', () => {
    const component = shallow(<PieChart data={emptyData} message="Hey" />);
    expect(component.contains('Hey')).toEqual(true);
  });

  it('renders `innerComponent` if provided', () => {
    const component = shallow(<PieChart data={data} innerComponent="Hey" />);
    expect(component.find('.pieChart__inner').text()).toEqual('Hey');
  });

  it('renders `label` and `labelPrefix` if provided', () => {
    const component = shallow(
      <PieChart data={data} labelPrefix={() => '$20'} label="redemptions" />
    );
    expect(
      component
        .find('.pieChart__label')
        .first()
        .text()
    ).toEqual('$20 redemptions');
  });

  it('renders provided `className`', () => {
    const component = shallow(<PieChart data={data} className="hey" />);
    expect(component.hasClass('hey')).toEqual(true);
  });

  it('stacks for small screens if `stackForSmall` is specified', () => {
    const component = shallow(<PieChart data={data} stackForSmall />);
    expect(
      component.find('.pieChart__legend').hasClass('small-12 medium-expand')
    ).toEqual(true);
  });

  it('renders specified `stroke`', () => {
    const component = shallow(<PieChart data={data} stroke="#000000" />);
    expect(
      component
        .find('Cell')
        .first()
        .prop('stroke')
    ).toEqual('#000000');
  });

  it('renders specified `radius` and `thickness`', () => {
    const component = shallow(
      <PieChart data={data} radius={50} thickness={10} />
    );
    expect(component.find('PieChart').prop('width')).toEqual(100);
    expect(component.find('PieChart').prop('height')).toEqual(100);
    expect(component.find('Pie').prop('innerRadius')).toEqual(38);
    expect(component.find('Pie').prop('outerRadius')).toEqual(50);
  });

  it('renders data if provided', () => {
    const component = create(<PieChart data={data} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
