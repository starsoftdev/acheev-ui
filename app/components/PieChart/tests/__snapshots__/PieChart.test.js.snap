// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PieChart /> renders data if provided 1`] = `
<div
  className="row align-middle align-center"
>
  <div
    className="shrink column"
  >
    <div
      className="pieChart"
    >
      <div
        className="recharts-wrapper pieChart__svg"
        style={
          Object {
            "cursor": "default",
            "height": 225,
            "position": "relative",
            "width": 225,
          }
        }
      >
        <svg
          className="recharts-surface"
          height={225}
          version="1.1"
          viewBox="0 0 225 225"
          width={225}
        >
          <clipPath
            id="recharts1-clip"
          >
            <rect
              height={215}
              width={215}
              x={5}
              y={5}
            />
          </clipPath>
          <g
            className="recharts-layer recharts-pie"
          >
            <g
              clipPath="url(#recharts-pie-2)"
            >
              <g
                className="recharts-layer"
              >
                <g
                  className="recharts-layer recharts-pie-sector"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                />
                <g
                  className="recharts-layer recharts-pie-sector"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                />
                <g
                  className="recharts-layer recharts-pie-sector"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                />
              </g>
            </g>
          </g>
        </svg>
      </div>
    </div>
  </div>
  <div
    className="pieChart__legend column"
  >
    <div
      className="pieChart__pancakeRow row align-middle"
    >
      <div
        className="shrink column"
      >
        <div
          style={
            Object {
              "width": 30,
            }
          }
        >
          <div
            className="pieChart__pancake"
            style={
              Object {
                "backgroundColor": "#003fa3",
                "height": 30,
                "width": 30,
              }
            }
          />
        </div>
      </div>
      <div
        className="column"
      >
        <div
          className="pieChart__value"
        >
          40
        </div>
        <div
          className="pieChart__label"
        >
           
        </div>
      </div>
    </div>
    <div
      className="pieChart__pancakeRow row align-middle"
    >
      <div
        className="shrink column"
      >
        <div
          style={
            Object {
              "width": 30,
            }
          }
        >
          <div
            className="pieChart__pancake"
            style={
              Object {
                "backgroundColor": "#017fcc",
                "height": 15,
                "width": 15,
              }
            }
          />
        </div>
      </div>
      <div
        className="column"
      >
        <div
          className="pieChart__value"
        >
          20
        </div>
        <div
          className="pieChart__label"
        >
           
        </div>
      </div>
    </div>
    <div
      className="pieChart__pancakeRow row align-middle"
    >
      <div
        className="shrink column"
      >
        <div
          style={
            Object {
              "width": 30,
            }
          }
        >
          <div
            className="pieChart__pancake"
            style={
              Object {
                "backgroundColor": "#11b4e4",
                "height": 15,
                "width": 15,
              }
            }
          />
        </div>
      </div>
      <div
        className="column"
      >
        <div
          className="pieChart__value"
        >
          20
        </div>
        <div
          className="pieChart__label"
        >
           
        </div>
      </div>
    </div>
  </div>
</div>
`;
